lib/Backend/FlowGraph.cpp

```diff
@@ -4434,6 +4434,312 @@ BasicBlock::CleanUpValueMaps()
     }
 }
 
+bool IsLegalOpcodeForPathDepBrFold(IR::Instr *instr)
+{
+    if (!instr->IsRealInstr())
+    {
+        return true;
+    }
+    switch (instr->m_opcode)
+    {
+    case Js::OpCode::Ld_A:
+    case Js::OpCode::Ld_I4:
+    case Js::OpCode::ByteCodeUses:
+        return true;
+    }
+#if DBG
+    if (PHASE_TRACE(Js::PathDepBranchFoldingPhase, instr->m_func) && Js::Configuration::Global.flags.Verbose)
+    {
+        Output::Print(_u("Skipping PathDependentBranchFolding due to: "));
+        instr->Dump();
+    }
+#endif
+    return false;
+}
+
+IR::LabelInstr* BasicBlock::CanProveConditionalBranch(IR::BranchInstr *branch, GlobOpt* globOpt, GlobHashTable * localSymToValueMap)
+{
+    if (!branch->GetSrc1() || !branch->GetSrc1()->GetStackSym())
+    {
+        return nullptr;
+    }
+
+    Value *src1Val = nullptr, *src2Val = nullptr;
+    Js::Var src1Var = nullptr, src2Var = nullptr;
+
+    Value ** src1ValPtr = localSymToValueMap->Get(branch->GetSrc1()->GetStackSym());
+    if (!src1ValPtr)
+    {
+        src1Val = this->globOptData.FindValue(branch->GetSrc1()->GetStackSym());
+        if (!src1Val)
+        {
+            return nullptr;
+        }
+    }
+    else
+    {
+        src1Val = *src1ValPtr;
+        if (!src1Val)
+        {
+            return nullptr;
+        }
+    }
+
+    src1Var = globOpt->GetConstantVar(branch->GetSrc1(), src1Val);
+
+    if (branch->GetSrc2() != nullptr)
+    {
+        if (branch->GetSrc2()->GetStackSym())
+        {
+            Value ** src2ValPtr = localSymToValueMap->Get(branch->GetSrc2()->GetStackSym());
+            if (!src2ValPtr)
+            {
+                src2Val = this->globOptData.FindValue(branch->GetSrc2()->GetStackSym());
+                if (!src2Val)
+                {
+                    return nullptr;
+                }
+            }
+            else
+            {
+                src2Val = *src2ValPtr;
+                if (!src2Val)
+                {
+                    return nullptr;
+                }
```

Comment:
nit: these checks are identical for src1 and src2. These should ideally be captured in a lambda or a separate method.